File: runtimes/runtime.js
const net = require('net');

if (process.stdout._handle) {
	process.stdout._handle.setBlocking(true);
}

function log(msg, data) {
	process.stdout.write(JSON.stringify({
		"level": "debug",
		"context": "runtime",
		"data": data,
		"msg": msg,
	}) + "\n")
}

function promisify(fn) {
	return function (...args) {
		return new Promise((resolve, reject) => {
			fn(...args, (err, result) => {
				if (err) {
					return reject(err);
				}
				resolve(result);
			});
		});
	};
}

const socket = process.env.PROTOFLOW_SOCKET;
if (!socket) {
	throw new Error("PROTOFLOW_SOCKET environment variable not set");
}

const sock = new net.createConnection({ path : socket });

sock.on('error', function(err) {
	console.error(err);
	flushOutputAndExit(-1);
});

sock.on('close', function() {
	flushOutputAndExit(0);
});

sock.on('data', function(data) {
	const inputData = JSON.parse(data.toString());
	log('received data', inputData);

	void init(inputData);
})

async function returnResult(socket, programOutput) {
	sock.write(JSON.stringify(programOutput));
	sock.end();
}

async function init(inputData) {
	try {
		// Inside of the try catch statement to ensure that any exceptions thrown here are logged.
		const socket = inputData[ "socket" ];
		const input = inputData[ "input" ];

		const importPath = inputData["import_path"];
		const functionName = inputData["function_name"];

		const importedFile = require(importPath);
		const mainEntrypoint = importedFile[functionName];

		const result = await mainEntrypoint(input);

		await returnResult(socket, {
			"result": result,
		});
	} catch ( e ) {
		log("error", e.toString());
		if( e.stack ) {
			e = e.stack.toString()
		} else {
			e = e.toString();
		}
		flushOutputAndExit(-1);
	}
}

function flushOutputAndExit(exitCode) {
	// Configure the streams to be blocking
	blockStreams(process.stdout);
	blockStreams(process.stderr);


	// Allow Node to cleanup any internals before the next process tick
	setImmediate(function callProcessExitWithCode() {
		process.exit(exitCode);
	});
}

function blockStreams(stream) {
	if (!stream || !stream._handle || !stream._handle.setBlocking) {
		// Not able to set blocking so just bail out
		return;
	}
	stream._handle.setBlocking(true);
}

File: runtimes/runtime.py
import asyncio
import json
import os
import sys
import traceback
import socket


def log(msg, data):
    content = json.dumps({
        "type": "log",
        "context": "runtime",
        "data": data,
        "msg": msg
    })
    print(content, file=sys.stderr)


def read_json_from_socket(s: socket.socket):
    data = s.recv(1024)
    return json.loads(data.decode("utf-8"))


def print_return_data(s, result, error=None):
    return_data = json.dumps({
        "result": result,
        "error": error
    })
    s.sendall(return_data.encode())


async def main():
    unix_socket = os.environ["PROTOFLOW_SOCKET"]
    if not unix_socket:
        log("PROTOFLOW_SOCKET not set", {})
        exit(-1)

    with socket.socket(socket.AF_UNIX, socket.SOCK_STREAM) as s:
        s.connect(unix_socket)

        input_data = read_json_from_socket(s)

        block_input = input_data["input"]

        import_path = input_data["import_path"]
        function_name = input_data["function_name"]

        try:
            imported_file = __import__(import_path)
            func = getattr(imported_file, function_name)

            result = await func(block_input)

            print_return_data(s, result)
        except Exception as e:
            log("Error running function", {
                "error": str(e),
                "traceback": traceback.format_exc()
            })
            print_return_data(s, None, traceback.format_exc())
            exit(-1)

        exit(0)


if __name__ == "__main__":
    asyncio.run(main())
